File: backend/src/main/domain/models/scientific_results.py
"""Pure domain contract for the scientific-results publication artifact (ADR 0025).

`scientific_results.json` is a sibling of `project_status.json` (ADR 0022/0023): a
separate, statically-published, git-tracked artifact carrying Nexus's empirical
outputs, never the system's engineering/audit status. This module defines only the
contract boundary — the schema and its invariants — not a publisher. No producer for
this contract exists yet; nothing in this module is wired to real Nexus output.

Track semantics (ADR 0025 §1-2):
- `discovery` (Head A, ADR 0017): landscape/candidate/adversarial synthesis output.
  Real-time, LLM-involved, not reproducible. Must never be presented as scientifically
  verified, and must carry the exact `DISCOVERY_DISCLAIMER` wherever rendered.
- `verification` (Head B, ADR 0017): deterministic, policy-sealed matching evidence.

Reuses existing payload shapes rather than redefining them: `PatentCluster`,
`InventionCandidate`, `AdversarialVerdict`, `ScoreCard` (domain.models.runtime_schemas,
Head A) and `MatchAssessment` (domain.models.matching, Head B). This module adds only
the publication envelope, execution scoping, evidence-citation structure, and the
epistemic invariants ADR 0025 requires around them — it does not create producers
for `OpportunityScore`/`OpportunityHypothesis` (domain.models.opportunity), which
remain unused by any pipeline.

`TechnologyClusterObservation` was added to carry the white-space metrics
(density/recency/citation_traction/demand_intensity/quadrant/mean_age_years) that
`application.landscape.metrics.compute_white_space_metrics` already computes but
`PatentCluster`/`cluster_patents` discard — this is the concrete publisher requirement
(the first real snapshot, see `scripts/publish_scientific_results.py`) that justified
extending this contract beyond PR #71's original scope.

Discovery-track free text is additionally checked against a fixed phrase denylist
(see `_FORBIDDEN_DISCOVERY_CLAIM_PHRASES` below). That check is a conservative
lexical guardrail against the most obvious mistakes reaching publication, not a
semantic claim verifier — it can both over- and under-reject; see its docstring.
"""

import re
from datetime import datetime
from enum import StrEnum
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

from domain.models.matching import MatchAssessment
from domain.models.runtime_schemas import AdversarialVerdict, InventionCandidate, PatentCluster, ScoreCard


class Track(StrEnum):
    """Mandatory, non-interchangeable classification of a published record (ADR 0025 §1)."""

    DISCOVERY = "discovery"
    VERIFICATION = "verification"


# Exact required wording (ADR 0025 §2). A discovery-track record's disclaimer must match
# this string verbatim — not a dashboard paraphrase (SCIENTIFIC_RESULTS_CONTRACT.md §5.8).
DISCOVERY_DISCLAIMER = (
    "Nexus generated/discovered this output; it is model-synthesized technology "
    "discovery output, not scientifically verified evidence. It does not demonstrate "
    "patentability, freedom-to-operate, or commercial efficacy, and is not a "
    "substitute for a professional prior-art search. (ADR 0017 §5, ADR 0025 §2)"
)

# CONSERVATIVE LEXICAL GUARDRAIL — not semantic/epistemic enforcement.
#
# This is a denylist substring match against a fixed phrase list, matched
# case-insensitively. It catches the obvious, unambiguous case: free text that
# asserts one of these exact phrases. It is deliberately NOT a claim-detection
# model and must not be treated as one:
#   - false positives are possible (e.g. text that *quotes* or *negates* a
#     forbidden phrase to explicitly disclaim it would still be rejected here);
#   - false negatives are possible (any rephrasing, translation, or claim not
#     on this exact list passes silently).
# Real enforcement of "discovery output must never imply patentability, FTO,
# efficacy, or scientific verification" (ADR 0017 §5.5-5.6, ADR 0025 §2) is a
# generation/review responsibility upstream of this contract — this guardrail
# exists only to fail loudly on the most obvious mistakes reaching publication,
# not to guarantee the invariant holds for everything that passes it.
_FORBIDDEN_DISCOVERY_CLAIM_PHRASES: tuple[str, ...] = (
    "is patentable",
    "is not patentable",
    "freedom to operate",
    "freedom-to-operate",
    "proven effective",
    "clinically proven",
    "scientifically verified",
    "scientifically proven",
    "demonstrates efficacy",
    "commercially viable",
    "guaranteed to",
)

_SHA256_RE = re.compile(r"^[0-9a-f]{64}$")


def _require_exact_discovery_disclaimer(v: str, record_kind: str) -> str:
    if v != DISCOVERY_DISCLAIMER:
        raise ValueError(
            f"a discovery {record_kind} record must carry the exact required epistemic "
            "disclaimer verbatim (ADR 0025 §2) — not a paraphrase or omission"
        )
    return v


def _reject_forbidden_discovery_claims(*texts: str | None) -> None:
    """Conservative lexical guardrail (see comment above) — not a semantic claim verifier."""
    for text in texts:
        lowered = (text or "").lower()
        for phrase in _FORBIDDEN_DISCOVERY_CLAIM_PHRASES:
            if phrase in lowered:
                raise ValueError(
                    f"discovery-track text must not assert '{phrase}' "
                    "(ADR 0017 §5.5-5.6, ADR 0025 §2) — conservative lexical "
                    "guardrail, not a semantic claim verifier"
                )


class CitationResolution(StrEnum):
    """Whether an evidence citation's referenced publication was resolved to metadata."""

    RESOLVED = "resolved"
    UNRESOLVED = "unresolved"


class EvidenceCitation(BaseModel):
    """A structured, traceable reference to a cited prior-art publication.

    A citation that cannot be resolved to a `PatentRecord`/`PatentDocument` is kept as
    an explicit `unresolved` entry (SCIENTIFIC_RESULTS_CONTRACT.md §5 invariant 4) —
    never silently dropped and never conflated with a resolved one.
    """

    model_config = ConfigDict(frozen=True)

    publication_id: str = Field(min_length=1)
    role: Literal["supports", "challenges"]
    resolution: CitationResolution
    title: str | None = None
    publication_date: str | None = None

    @model_validator(mode="after")
    def validate_resolution_consistency(self) -> "EvidenceCitation":
        if self.resolution == CitationResolution.RESOLVED and self.title is None:
            raise ValueError("a 'resolved' citation must carry a resolved title")
        if self.resolution == CitationResolution.UNRESOLVED and (
            self.title is not None or self.publication_date is not None
        ):
            raise ValueError(
                "an 'unresolved' citation must not carry resolved metadata "
                "(title/publication_date) — that would misrepresent it as resolved"
            )
        return self


class ChallengeStatus(StrEnum):
    """Distinguishes 'no challenge has run yet' from 'a challenge ran' (SCIENTIFIC_RESULTS_CONTRACT.md
    §5 invariant 5: absence of evidence must never be represented as negative evidence).
    """

    NOT_YET_CHALLENGED = "not_yet_challenged"
    CHALLENGED = "challenged"


def _validate_utc_aware(v: datetime) -> datetime:
    if v.tzinfo is None:
        raise ValueError("timestamp must be timezone-aware (UTC)")
    return v


class ScientificResultsExecution(BaseModel):
    """Identity and provenance of one execution that produced published records.

    `dataset_id`/`dataset_version`/`policy_id`/`policy_version`/`policy_sha256`/
    `engine_commit` are optional: the deterministic (`verification`) track already
    stamps all of them (mirrors `EvaluationRunReport`); the `discovery` track does not
    yet (ADR 0017 §7 — Head A has no durable, versioned execution record today). Making
    them optional here reflects that gap honestly rather than fabricating values; see
    `executions_are_aggregable` below for why an absent field blocks aggregation rather
    than being treated as a wildcard match.
    """

    model_config = ConfigDict(frozen=True)

    execution_id: str = Field(min_length=1)
    track: Track
    domain: str = Field(min_length=1)
    query: str = Field(min_length=1)
    created_at: datetime
    dataset_id: str | None = None
    dataset_version: str | None = None
    engine_commit: str | None = None
    policy_id: str | None = None
    policy_version: str | None = None
    policy_sha256: str | None = None

    @field_validator("created_at")
    @classmethod
    def validate_created_at_tz_aware(cls, v: datetime) -> datetime:
        return _validate_utc_aware(v)

    @field_validator("policy_sha256")
    @classmethod
    def validate_policy_sha256_format(cls, v: str | None) -> str | None:
        if v is not None and not _SHA256_RE.match(v.lower()):
            raise ValueError(f"Invalid SHA-256 digest format: {v}")
        return v.lower() if v is not None else v


def executions_are_aggregable(a: ScientificResultsExecution, b: ScientificResultsExecution) -> bool:
    """Whether two executions may be combined into one aggregate figure.

    Two executions are aggregable only if they share `track`, `dataset_id`, and
    `policy_version`, and none of those fields is missing on either side — a missing
    field never matches, on either side, even against another missing field
    (SCIENTIFIC_RESULTS_CONTRACT.md §5 invariant 6). This is the single reusable rule a
    future dashboard/publisher must call before computing any cross-execution figure;
    it does not itself perform aggregation.
    """
    if a.track != b.track:
        return False
    if a.dataset_id is None or b.dataset_id is None or a.dataset_id != b.dataset_id:
        return False
    return not (a.policy_version is None or b.policy_version is None or a.policy_version != b.policy_version)


class DiscoveryCandidateRecord(BaseModel):
    """A published `discovery`-track candidate invention, scoped to its execution."""

    model_config = ConfigDict(frozen=True)

    execution_id: str = Field(min_length=1)
    candidate: InventionCandidate
    disclaimer: str = Field(default=DISCOVERY_DISCLAIMER)

    @field_validator("disclaimer")
    @classmethod
    def validate_disclaimer_is_exact(cls, v: str) -> str:
        return _require_exact_discovery_disclaimer(v, "candidate")

    @model_validator(mode="after")
    def validate_no_forbidden_claims(self) -> "DiscoveryCandidateRecord":
        _reject_forbidden_discovery_claims(
            self.candidate.title,
            self.candidate.description,
            self.candidate.claimed_novelty,
        )
        return self


class DiscoveryVerificationRecord(BaseModel):
    """A published `discovery`-track adversarial-challenge outcome for one candidate.

    `challenge_status` is the field that makes "not yet challenged" observably distinct
    from "challenged and found clean" — an `AdversarialVerdict` is never present without
    a `CHALLENGED` status, and never absent while still claiming `CHALLENGED`.
    """

    model_config = ConfigDict(frozen=True)

    execution_id: str = Field(min_length=1)
    candidate_id: str = Field(min_length=1)
    challenge_status: ChallengeStatus
    verdict: AdversarialVerdict | None = None
    scorecard: ScoreCard | None = None
    citations: tuple[EvidenceCitation, ...] = Field(default_factory=tuple)
    disclaimer: str = Field(default=DISCOVERY_DISCLAIMER)

    @field_validator("disclaimer")
    @classmethod
    def validate_disclaimer_is_exact(cls, v: str) -> str:
        return _require_exact_discovery_disclaimer(v, "verification")

    @model_validator(mode="after")
    def validate_challenge_consistency(self) -> "DiscoveryVerificationRecord":
        if self.challenge_status == ChallengeStatus.NOT_YET_CHALLENGED:
            if self.verdict is not None or self.citations:
                raise ValueError(
                    "a 'not_yet_challenged' record must carry no verdict and no "
                    "citations — an empty verdict/citation set on a challenge that "
                    "never ran would be indistinguishable from a completed challenge "
                    "that found nothing, which is exactly the ambiguity this status "
                    "exists to prevent"
                )
        elif self.verdict is None:
            raise ValueError("a 'challenged' record must carry its verdict")

        if self.verdict is not None:
            _reject_forbidden_discovery_claims(self.verdict.rationale)
        if self.scorecard is not None:
            _reject_forbidden_discovery_claims(*self.scorecard.supporting_evidence)
        return self


class VerificationMatchRecord(BaseModel):
    """A published `verification`-track demand-patent match assessment, scoped to its execution.

    Carries no disclaimer: `verification`-track evidence is the deterministic, policy-sealed
    track ADR 0025 §2 permits describing as "evaluated under this deterministic policy."
    """

    model_config = ConfigDict(frozen=True)

    execution_id: str = Field(min_length=1)
    assessment: MatchAssessment


class TechnologyClusterObservation(BaseModel):
    """A published cluster observation: the existing `PatentCluster` plus the fuller
    white-space metrics `compute_white_space_metrics`
    (application.landscape.metrics) already computes but `cluster_patents`
    (application.landscape.clustering) discards before returning. No new computation
    is introduced by this type — every field is copied verbatim from data Nexus's
    existing landscape pipeline already produces (SCIENTIFIC_RESULTS_CONTRACT.md §4,
    `TechnologyCluster` "Derivable" row).
    """

    model_config = ConfigDict(frozen=True)

    cluster: PatentCluster
    density: float
    recency: float
    citation_traction: float
    citation_coverage: float
    demand_intensity: float
    quadrant: str = Field(min_length=1)
    mean_age_years: float


class DiscoveryLandscapeRecord(BaseModel):
    """A published `discovery`-track landscape observation, scoped to its execution."""

    model_config = ConfigDict(frozen=True)

    execution_id: str = Field(min_length=1)
    query: str = Field(min_length=1)
    clusters: tuple[TechnologyClusterObservation, ...] = Field(default_factory=tuple)
    disclaimer: str = Field(default=DISCOVERY_DISCLAIMER)

    @field_validator("disclaimer")
    @classmethod
    def validate_disclaimer_is_exact(cls, v: str) -> str:
        return _require_exact_discovery_disclaimer(v, "landscape")


class ScientificResultsDocument(BaseModel):
    """Top-level `scientific_results.json` contract.

    Enforces referential integrity between records and their declared execution
    (every record must reference a real, listed execution) and track consistency
    (a record's track must match the track of the execution it references) —
    the mechanism that makes "discovery and verification cannot be silently mixed"
    a structural guarantee rather than a convention.
    """

    model_config = ConfigDict(frozen=True)

    schema_version: str = Field(min_length=1)
    generated_at: datetime
    executions: tuple[ScientificResultsExecution, ...] = Field(default_factory=tuple)
    landscapes: tuple[DiscoveryLandscapeRecord, ...] = Field(default_factory=tuple)
    candidates: tuple[DiscoveryCandidateRecord, ...] = Field(default_factory=tuple)
    verifications: tuple[DiscoveryVerificationRecord, ...] = Field(default_factory=tuple)
    matches: tuple[VerificationMatchRecord, ...] = Field(default_factory=tuple)

    @field_validator("generated_at")
    @classmethod
    def validate_generated_at_tz_aware(cls, v: datetime) -> datetime:
        return _validate_utc_aware(v)

    @model_validator(mode="after")
    def validate_referential_integrity(self) -> "ScientificResultsDocument":
        executions_by_id: dict[str, ScientificResultsExecution] = {}
        for execution in self.executions:
            if execution.execution_id in executions_by_id:
                raise ValueError(f"duplicate execution_id in executions: '{execution.execution_id}'")
            executions_by_id[execution.execution_id] = execution

        for landscape in self.landscapes:
            self._require_track(landscape.execution_id, executions_by_id, Track.DISCOVERY, "landscape")
        for candidate in self.candidates:
            self._require_track(candidate.execution_id, executions_by_id, Track.DISCOVERY, "candidate")
        for verification in self.verifications:
            self._require_track(verification.execution_id, executions_by_id, Track.DISCOVERY, "verification")
        for match in self.matches:
            self._require_track(match.execution_id, executions_by_id, Track.VERIFICATION, "match")
        return self

    @staticmethod
    def _require_track(
        execution_id: str,
        executions_by_id: dict[str, ScientificResultsExecution],
        required_track: Track,
        record_kind: str,
    ) -> None:
        execution = executions_by_id.get(execution_id)
        if execution is None:
            raise ValueError(f"{record_kind} references unknown execution_id '{execution_id}'")
        if execution.track != required_track:
            raise ValueError(
                f"{record_kind} references execution_id '{execution_id}' with "
                f"track={execution.track}, but a {record_kind} record requires "
                f"track={required_track} — discovery and verification cannot be mixed "
                "(ADR 0025 §1)"
            )

File: nexus-status/frontend/src/domain/scientificResults.ts
/**
 * Scientific Results Contract v1 Domain Model (ADR 0025).
 *
 * Pure domain model representing the canonical `scientific_results.json` contract —
 * a sibling artifact to `project_status.json` (ADR 0022/0023), never merged with it:
 * this module carries Nexus's empirical outputs, not its engineering/audit status.
 *
 * Invariants:
 * 1. Mandatory, non-interchangeable track vocabulary: `discovery` (Head A) vs
 *    `verification` (Head B) (ADR 0025 §1). A record's track must match the track of
 *    the execution it references — discovery and verification are never silently mixed.
 * 2. Immutable, read-only representation of published evidence.
 * 3. Epistemic restraint: no scientific value (score, metric, verdict) is
 *    recalculated, derived, or synthesized here — every value is copied verbatim from
 *    the published artifact.
 * 4. Absent optional provenance (dataset_id, policy_id, etc.) is preserved as `null`,
 *    never defaulted or fabricated.
 * 5. Zero runtime UI or framework dependencies (pure TypeScript).
 */

export const TRACK_VALUES = ['discovery', 'verification'] as const;
export type Track = (typeof TRACK_VALUES)[number];

export const CHALLENGE_STATUS_VALUES = ['not_yet_challenged', 'challenged'] as const;
export type ChallengeStatus = (typeof CHALLENGE_STATUS_VALUES)[number];

export const CITATION_RESOLUTION_VALUES = ['resolved', 'unresolved'] as const;
export type CitationResolution = (typeof CITATION_RESOLUTION_VALUES)[number];

export const CITATION_ROLE_VALUES = ['supports', 'challenges'] as const;
export type CitationRole = (typeof CITATION_ROLE_VALUES)[number];

export const SOURCE_PROVENANCE_ROLE_VALUES = ['patents', 'demand'] as const;
export type SourceProvenanceRole = (typeof SOURCE_PROVENANCE_ROLE_VALUES)[number];

export class ScientificResultsValidationError extends Error {
  constructor(message: string) {
    super(`ScientificResultsValidationError: ${message}`);
    this.name = 'ScientificResultsValidationError';
  }
}

export interface DataSourceProvenance {
  readonly role: SourceProvenanceRole;
  readonly kind: string;
}

export interface ScientificResultsExecution {
  readonly execution_id: string;
  readonly track: Track;
  readonly domain: string;
  readonly query: string;
  readonly created_at: string;
  readonly dataset_id: string | null;
  readonly dataset_version: string | null;
  readonly engine_commit: string | null;
  readonly policy_id: string | null;
  readonly policy_version: string | null;
  readonly policy_sha256: string | null;
  readonly source_provenance: readonly DataSourceProvenance[];
}

export interface PatentCluster {
  readonly cluster_id: string;
  readonly label: string;
  readonly representative_patents: readonly string[];
  readonly patent_count: number;
  readonly white_space_score: number;
  readonly is_white_space: boolean;
}

export interface TechnologyClusterObservation {
  readonly cluster: PatentCluster;
  readonly density: number;
  readonly recency: number;
  readonly citation_traction: number;
  readonly citation_coverage: number;
  readonly demand_intensity: number;
  readonly quadrant: string;
  readonly mean_age_years: number;
}

export interface DiscoveryLandscapeRecord {
  readonly execution_id: string;
  readonly query: string;
  readonly clusters: readonly TechnologyClusterObservation[];
  readonly disclaimer: string;
}

export interface InventionCandidate {
  readonly candidate_id: string;
  readonly cluster_id: string;
  readonly title: string;
  readonly description: string;
  readonly claimed_novelty: string;
}

export interface DiscoveryCandidateRecord {
  readonly execution_id: string;
  readonly candidate: InventionCandidate;
  readonly disclaimer: string;
}

export interface EvidenceCitation {
  readonly publication_id: string;
  readonly role: CitationRole;
  readonly resolution: CitationResolution;
  readonly title: string | null;
  readonly publication_date: string | null;
}

export interface AdversarialVerdict {
  readonly candidate_id: string;
  readonly verdict: string;
  readonly rationale: string;
  readonly cited_patents: readonly string[];
}

export interface ScoreCard {
  readonly candidate_id: string;
  readonly novelty: number;
  readonly prior_art_risk: number;
  readonly differentiation: number;
  readonly evidence: number;
  readonly supporting_evidence: readonly string[];
}

export interface DiscoveryVerificationRecord {
  readonly execution_id: string;
  readonly candidate_id: string;
  readonly challenge_status: ChallengeStatus;
  readonly verdict: AdversarialVerdict | null;
  readonly scorecard: ScoreCard | null;
  readonly citations: readonly EvidenceCitation[];
  readonly disclaimer: string;
}

export interface MatchAssessment {
  readonly demand_id: string;
  readonly publication_id: string;
  readonly overall_score: number;
  readonly confidence: string;
  readonly sufficiency: string;
  readonly rationale: string;
  readonly policy_id: string;
  readonly policy_version: string;
  readonly policy_sha256: string;
  readonly fusion_transform_id: string;
}

export interface VerificationMatchRecord {
  readonly execution_id: string;
  readonly assessment: MatchAssessment;
}

export interface ScientificResultsDocument {
  readonly schema_version: string;
  readonly generated_at: string;
  readonly executions: readonly ScientificResultsExecution[];
  readonly landscapes: readonly DiscoveryLandscapeRecord[];
  readonly candidates: readonly DiscoveryCandidateRecord[];
  readonly verifications: readonly DiscoveryVerificationRecord[];
  readonly matches: readonly VerificationMatchRecord[];
}

const ISO_8601_PATTERN = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d+)?(?:Z|[+-]\d{2}:\d{2})$/;

function isIsoTimestamp(value: unknown): value is string {
  return typeof value === 'string' && ISO_8601_PATTERN.test(value) && !Number.isNaN(Date.parse(value));
}

function isTrack(value: unknown): value is Track {
  return typeof value === 'string' && (TRACK_VALUES as readonly string[]).includes(value);
}

function asRecord(value: unknown, context: string): Record<string, unknown> {
  if (typeof value !== 'object' || value === null || Array.isArray(value)) {
    throw new ScientificResultsValidationError(`${context} must be an object.`);
  }
  return value as Record<string, unknown>;
}

function asArray(value: unknown, context: string): unknown[] {
  if (!Array.isArray(value)) {
    throw new ScientificResultsValidationError(`${context} must be an array.`);
  }
  return value;
}

function requireString(value: unknown, context: string, allowEmpty = false): string {
  if (typeof value !== 'string' || (!allowEmpty && value.trim() === '')) {
    throw new ScientificResultsValidationError(`${context} must be a non-empty string.`);
  }
  return value;
}

function requireNumber(value: unknown, context: string): number {
  if (typeof value !== 'number' || Number.isNaN(value)) {
    throw new ScientificResultsValidationError(`${context} must be a number.`);
  }
  return value;
}

function requireBoolean(value: unknown, context: string): boolean {
  if (typeof value !== 'boolean') {
    throw new ScientificResultsValidationError(`${context} must be a boolean.`);
  }
  return value;
}

function optionalString(value: unknown, context: string): string | null {
  if (value === null || value === undefined) return null;
  if (typeof value !== 'string') {
    throw new ScientificResultsValidationError(`${context} must be a string or null.`);
  }
  return value;
}

function requireStringArray(value: unknown, context: string): readonly string[] {
  return Object.freeze(
    asArray(value, context).map((item, i) => requireString(item, `${context}[${i}]`))
  );
}

function validateDataSourceProvenance(raw: unknown, index: number): DataSourceProvenance {
  const p = asRecord(raw, `source_provenance[${index}]`);
  const role = p.role;
  if (role !== 'patents' && role !== 'demand') {
    throw new ScientificResultsValidationError(
      `source_provenance[${index}].role must be 'patents' or 'demand', got '${String(role)}'.`
    );
  }
  return Object.freeze({
    role,
    kind: requireString(p.kind, `source_provenance[${index}].kind`),
  });
}

function validateExecution(raw: unknown, index: number): ScientificResultsExecution {
  const e = asRecord(raw, `executions[${index}]`);
  const executionId = requireString(e.execution_id, `executions[${index}].execution_id`);

  if (!isTrack(e.track)) {
    throw new ScientificResultsValidationError(
      `executions[${index}].track must be one of: ${TRACK_VALUES.join(', ')}. Got '${String(e.track)}'.`
    );
  }
  if (!isIsoTimestamp(e.created_at)) {
    throw new ScientificResultsValidationError(
      `executions[${index}].created_at must be an ISO-8601 timestamp. Got '${String(e.created_at)}'.`
    );
  }

  const sourceProvenance = e.source_provenance === undefined
    ? []
    : asArray(e.source_provenance, `executions[${index}].source_provenance`).map((sp, i) =>
        validateDataSourceProvenance(sp, i)
      );

  return Object.freeze({
    execution_id: executionId,
    track: e.track,
    domain: requireString(e.domain, `executions[${index}].domain`),
    query: requireString(e.query, `executions[${index}].query`),
    created_at: e.created_at,
    dataset_id: optionalString(e.dataset_id, `executions[${index}].dataset_id`),
    dataset_version: optionalString(e.dataset_version, `executions[${index}].dataset_version`),
    engine_commit: optionalString(e.engine_commit, `executions[${index}].engine_commit`),
    policy_id: optionalString(e.policy_id, `executions[${index}].policy_id`),
    policy_version: optionalString(e.policy_version, `executions[${index}].policy_version`),
    policy_sha256: optionalString(e.policy_sha256, `executions[${index}].policy_sha256`),
    source_provenance: Object.freeze(sourceProvenance),
  });
}

function validatePatentCluster(raw: unknown, context: string): PatentCluster {
  const c = asRecord(raw, context);
  return Object.freeze({
    cluster_id: requireString(c.cluster_id, `${context}.cluster_id`),
    label: requireString(c.label, `${context}.label`),
    representative_patents: requireStringArray(c.representative_patents, `${context}.representative_patents`),
    patent_count: requireNumber(c.patent_count, `${context}.patent_count`),
    white_space_score: requireNumber(c.white_space_score, `${context}.white_space_score`),
    is_white_space: requireBoolean(c.is_white_space, `${context}.is_white_space`),
  });
}

function validateClusterObservation(raw: unknown, index: number, context: string): TechnologyClusterObservation {
  const o = asRecord(raw, `${context}.clusters[${index}]`);
  return Object.freeze({
    cluster: validatePatentCluster(o.cluster, `${context}.clusters[${index}].cluster`),
    density: requireNumber(o.density, `${context}.clusters[${index}].density`),
    recency: requireNumber(o.recency, `${context}.clusters[${index}].recency`),
    citation_traction: requireNumber(o.citation_traction, `${context}.clusters[${index}].citation_traction`),
    citation_coverage: requireNumber(o.citation_coverage, `${context}.clusters[${index}].citation_coverage`),
    demand_intensity: requireNumber(o.demand_intensity, `${context}.clusters[${index}].demand_intensity`),
    quadrant: requireString(o.quadrant, `${context}.clusters[${index}].quadrant`),
    mean_age_years: requireNumber(o.mean_age_years, `${context}.clusters[${index}].mean_age_years`),
  });
}

function validateLandscape(raw: unknown, index: number): DiscoveryLandscapeRecord {
  const context = `landscapes[${index}]`;
  const l = asRecord(raw, context);
  const clusters = asArray(l.clusters ?? [], `${context}.clusters`).map((c, i) =>
    validateClusterObservation(c, i, context)
  );
  return Object.freeze({
    execution_id: requireString(l.execution_id, `${context}.execution_id`),
    query: requireString(l.query, `${context}.query`),
    clusters: Object.freeze(clusters),
    disclaimer: requireString(l.disclaimer, `${context}.disclaimer`),
  });
}

function validateCandidate(raw: unknown, index: number): DiscoveryCandidateRecord {
  const context = `candidates[${index}]`;
  const r = asRecord(raw, context);
  const c = asRecord(r.candidate, `${context}.candidate`);
  return Object.freeze({
    execution_id: requireString(r.execution_id, `${context}.execution_id`),
    candidate: Object.freeze({
      candidate_id: requireString(c.candidate_id, `${context}.candidate.candidate_id`),
      cluster_id: requireString(c.cluster_id, `${context}.candidate.cluster_id`),
      title: requireString(c.title, `${context}.candidate.title`),
      description: requireString(c.description, `${context}.candidate.description`),
      claimed_novelty: requireString(c.claimed_novelty, `${context}.candidate.claimed_novelty`, true),
    }),
    disclaimer: requireString(r.disclaimer, `${context}.disclaimer`),
  });
}

function validateCitation(raw: unknown, index: number, context: string): EvidenceCitation {
  const c = asRecord(raw, `${context}.citations[${index}]`);
  const role = c.role;
  if (role !== 'supports' && role !== 'challenges') {
    throw new ScientificResultsValidationError(
      `${context}.citations[${index}].role must be 'supports' or 'challenges'. Got '${String(role)}'.`
    );
  }
  const resolution = c.resolution;
  if (resolution !== 'resolved' && resolution !== 'unresolved') {
    throw new ScientificResultsValidationError(
      `${context}.citations[${index}].resolution must be 'resolved' or 'unresolved'. Got '${String(resolution)}'.`
    );
  }
  return Object.freeze({
    publication_id: requireString(c.publication_id, `${context}.citations[${index}].publication_id`),
    role,
    resolution,
    title: optionalString(c.title, `${context}.citations[${index}].title`),
    publication_date: optionalString(c.publication_date, `${context}.citations[${index}].publication_date`),
  });
}

function validateVerdict(raw: unknown, context: string): AdversarialVerdict {
  const v = asRecord(raw, context);
  return Object.freeze({
    candidate_id: requireString(v.candidate_id, `${context}.candidate_id`),
    verdict: requireString(v.verdict, `${context}.verdict`),
    rationale: requireString(v.rationale, `${context}.rationale`, true),
    cited_patents: requireStringArray(v.cited_patents, `${context}.cited_patents`),
  });
}

function validateScoreCard(raw: unknown, context: string): ScoreCard {
  const s = asRecord(raw, context);
  return Object.freeze({
    candidate_id: requireString(s.candidate_id, `${context}.candidate_id`),
    novelty: requireNumber(s.novelty, `${context}.novelty`),
    prior_art_risk: requireNumber(s.prior_art_risk, `${context}.prior_art_risk`),
    differentiation: requireNumber(s.differentiation, `${context}.differentiation`),
    evidence: requireNumber(s.evidence, `${context}.evidence`),
    supporting_evidence: requireStringArray(s.supporting_evidence, `${context}.supporting_evidence`),
  });
}

function validateVerification(raw: unknown, index: number): DiscoveryVerificationRecord {
  const context = `verifications[${index}]`;
  const r = asRecord(raw, context);
  const challengeStatus = r.challenge_status;
  if (!(CHALLENGE_STATUS_VALUES as readonly string[]).includes(challengeStatus as string)) {
    throw new ScientificResultsValidationError(
      `${context}.challenge_status must be one of: ${CHALLENGE_STATUS_VALUES.join(', ')}. Got '${String(challengeStatus)}'.`
    );
  }
  const citations = asArray(r.citations ?? [], `${context}.citations`).map((c, i) =>
    validateCitation(c, i, context)
  );
  return Object.freeze({
    execution_id: requireString(r.execution_id, `${context}.execution_id`),
    candidate_id: requireString(r.candidate_id, `${context}.candidate_id`),
    challenge_status: challengeStatus as ChallengeStatus,
    verdict: r.verdict == null ? null : validateVerdict(r.verdict, `${context}.verdict`),
    scorecard: r.scorecard == null ? null : validateScoreCard(r.scorecard, `${context}.scorecard`),
    citations: Object.freeze(citations),
    disclaimer: requireString(r.disclaimer, `${context}.disclaimer`),
  });
}

function validateMatch(raw: unknown, index: number): VerificationMatchRecord {
  const context = `matches[${index}]`;
  const r = asRecord(raw, context);
  const a = asRecord(r.assessment, `${context}.assessment`);
  return Object.freeze({
    execution_id: requireString(r.execution_id, `${context}.execution_id`),
    assessment: Object.freeze({
      demand_id: requireString(a.demand_id, `${context}.assessment.demand_id`),
      publication_id: requireString(a.publication_id, `${context}.assessment.publication_id`),
      overall_score: requireNumber(a.overall_score, `${context}.assessment.overall_score`),
      confidence: requireString(a.confidence, `${context}.assessment.confidence`),
      sufficiency: requireString(a.sufficiency, `${context}.assessment.sufficiency`),
      rationale: requireString(a.rationale, `${context}.assessment.rationale`, true),
      policy_id: requireString(a.policy_id, `${context}.assessment.policy_id`),
      policy_version: requireString(a.policy_version, `${context}.assessment.policy_version`),
      policy_sha256: requireString(a.policy_sha256, `${context}.assessment.policy_sha256`),
      fusion_transform_id: requireString(a.fusion_transform_id, `${context}.assessment.fusion_transform_id`),
    }),
  });
}

/**
 * Parses and validates an untrusted raw payload against the Scientific Results
 * Contract v1 (ADR 0025). Throws ScientificResultsValidationError on any structural
 * or referential-integrity violation — never silently downgrades a corrupt/invalid
 * payload into an empty result set.
 *
 * Referential integrity (ADR 0025 §1): every landscape/candidate/verification record
 * must reference a listed execution whose track is `discovery`; every match record
 * must reference a listed execution whose track is `verification`. This is the
 * mechanism that makes "discovery and verification cannot be mixed" structural on
 * the read side, mirroring the same invariant already enforced at publish time by
 * `ScientificResultsDocument` (backend/src/main/domain/models/scientific_results.py).
 */
export function parseScientificResults(raw: unknown): ScientificResultsDocument {
  const payload = asRecord(raw, 'Payload');

  const schemaVersion = requireString(payload.schema_version, 'schema_version');
  const generatedAt = payload.generated_at;
  if (!isIsoTimestamp(generatedAt)) {
    throw new ScientificResultsValidationError(
      `Invalid generated_at timestamp: '${String(generatedAt)}'. Must be ISO-8601.`
    );
  }

  const executions = asArray(payload.executions ?? [], 'executions').map((e, i) => validateExecution(e, i));
  const executionsById = new Map<string, ScientificResultsExecution>();
  for (const execution of executions) {
    if (executionsById.has(execution.execution_id)) {
      throw new ScientificResultsValidationError(`Duplicate execution_id in executions: '${execution.execution_id}'.`);
    }
    executionsById.set(execution.execution_id, execution);
  }

  function requireTrack(executionId: string, requiredTrack: Track, recordKind: string): void {
    const execution = executionsById.get(executionId);
    if (!execution) {
      throw new ScientificResultsValidationError(`${recordKind} references unknown execution_id '${executionId}'.`);
    }
    if (execution.track !== requiredTrack) {
      throw new ScientificResultsValidationError(
        `${recordKind} references execution_id '${executionId}' with track='${execution.track}', ` +
          `but a ${recordKind} record requires track='${requiredTrack}' — discovery and verification cannot be mixed (ADR 0025 §1).`
      );
    }
  }

  const landscapes = asArray(payload.landscapes ?? [], 'landscapes').map((l, i) => validateLandscape(l, i));
  landscapes.forEach((l) => requireTrack(l.execution_id, 'discovery', 'landscape'));

  const candidates = asArray(payload.candidates ?? [], 'candidates').map((c, i) => validateCandidate(c, i));
  candidates.forEach((c) => requireTrack(c.execution_id, 'discovery', 'candidate'));

  const verifications = asArray(payload.verifications ?? [], 'verifications').map((v, i) => validateVerification(v, i));
  verifications.forEach((v) => requireTrack(v.execution_id, 'discovery', 'verification'));

  const matches = asArray(payload.matches ?? [], 'matches').map((m, i) => validateMatch(m, i));
  matches.forEach((m) => requireTrack(m.execution_id, 'verification', 'match'));

  return Object.freeze({
    schema_version: schemaVersion,
    generated_at: generatedAt,
    executions: Object.freeze(executions),
    landscapes: Object.freeze(landscapes),
    candidates: Object.freeze(candidates),
    verifications: Object.freeze(verifications),
    matches: Object.freeze(matches),
  }) as ScientificResultsDocument;
}
